### PR TITLE
WIP: Implement moveit_cpp (reopen #1)

### PR DESCRIPTION
**Pull request by henningkayser**
I addressed most low-hanging fruits from Simon's PR [#1](https://github.com/PickNikRobotics/moveit/pull/1) and since I can't commit to his branch I reopened this one. This is WIP and I will probably refactor the entire class.
@davetcoleman I didn't answer some of your requested changes. I fully agree with those that are still open and will address them during the refactoring process.

My plan is to get this implementation functional for planning and execution (of course without actions/services), write a test and clean everything up. After that, I would continue adding features in follow-up PRs.

### CODE REVIEW

**Comment by henningkayser** [comment]
> Thanks for helping in improving MoveIt

That's great :+1:

File: moveit_ros/planning_interface/moveit_cpp/include/moveit/moveit_cpp/moveit_cpp.h [hunk 1]
```diff
@@ -0,0 +1,1101 @@
...
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
+
+/* Author: Simon Goldstein, Henning Kayser */
```
Comment (davetcoleman) [inline moveit_ros/planning_interface/moveit_cpp/include/moveit/moveit_cpp/moveit_cpp.h]: Henning I think you can claim sole authorship here... Simon's PR was basically a copy of the original?

File: moveit_ros/planning_interface/moveit_cpp/include/moveit/moveit_cpp/moveit_cpp.h [hunk 2]
```diff
@@ -0,0 +1,1101 @@
...
+#include <moveit_msgs/PlaceAction.h>
+#include <geometry_msgs/PoseStamped.h>
+#include <moveit/robot_state/conversions.h>
+#include <actionlib/client/simple_action_client.h>
+#include <memory>
+#include <tf2_ros/buffer.h>
+
+namespace planning_scene_monitor
+{
+MOVEIT_CLASS_FORWARD(PlanningSceneMonitor);
```
Comment (davetcoleman) [inline moveit_ros/planning_interface/moveit_cpp/include/moveit/moveit_cpp/moveit_cpp.h]: remove, the include is above

File: moveit_ros/planning_interface/moveit_cpp/include/moveit/moveit_cpp/moveit_cpp.h [hunk 3]
```diff
@@ -0,0 +1,1101 @@
...
+  }
+  bool operator!=(const int c) const
+  {
+    return val != c;
+  }
+};
+
+MOVEIT_CLASS_FORWARD(MoveItCpp);
+
+/** \class MoveItCpp move_group_interface.h moveit/planning_interface/move_group_interface.h
```
Comment (davetcoleman) [inline moveit_ros/planning_interface/moveit_cpp/include/moveit/moveit_cpp/moveit_cpp.h]: remove these wrong file names

File: moveit_ros/planning_interface/moveit_cpp/include/moveit/moveit_cpp/moveit_cpp.h [hunk 4]
```diff
@@ -0,0 +1,1101 @@
...
+
+    \brief Client class to conveniently use the ROS interfaces provided by the move_group node.
```
Comment (davetcoleman) [inline moveit_ros/planning_interface/moveit_cpp/include/moveit/moveit_cpp/moveit_cpp.h]: update with description from our specification gdoc

File: moveit_ros/planning_interface/moveit_cpp/include/moveit/moveit_cpp/moveit_cpp.h [hunk 5]
```diff
@@ -0,0 +1,1101 @@
...
+
+    /// The amount of time it took to generate the plan
+    double planning_time_;
+  };
+
+  /**
+      \brief Construct a MoveItCpp instance call using a specified set of options \e opt.
+
+      \param opt. A MoveItCpp::Options structure, if you pass a ros::NodeHandle with a specific callback queue,
+     it has to be of type ros::CallbackQueue
```
Comment (davetcoleman) [inline moveit_ros/planning_interface/moveit_cpp/include/moveit/moveit_cpp/moveit_cpp.h]: cleanup line wrapping

File: moveit_ros/planning_interface/moveit_cpp/include/moveit/moveit_cpp/moveit_cpp.h [hunk 6]
```diff
@@ -0,0 +1,1101 @@
...
+        (which is the default type of callback queues used in ROS)
+      \param tf_buffer. Specify a TF2_ROS Buffer instance to use. If not specified,
+                        one will be constructed internally along with an internal TF2_ROS TransformListener
+      \param wait_for_servers. Timeout for connecting to action servers. Zero time means unlimited waiting.
+    */
+  MoveItCpp(const Options& opt, const std::shared_ptr<tf2_ros::Buffer>& tf_buffer = std::shared_ptr<tf2_ros::Buffer>(),
+            const ros::WallDuration& wait_for_servers = ros::WallDuration());
+  /**
+      \brief Construct a client for the MoveGroup action for a particular \e group.
```
Comment (davetcoleman) [inline moveit_ros/planning_interface/moveit_cpp/include/moveit/moveit_cpp/moveit_cpp.h]: reword

File: moveit_ros/planning_interface/moveit_cpp/src/moveit.cpp [hunk 7]
```diff
@@ -0,0 +1,1912 @@
...
+      planning_scene_monitor::PlanningSceneMonitor::DEFAULT_ATTACHED_COLLISION_OBJECT_TOPIC, 1, false);
+
+  current_state_monitor_ = getSharedStateMonitor(robot_model_, tf_buffer_, node_handle_);
+
+  ros::WallTime timeout_for_servers = ros::WallTime::now() + wait_for_servers;
+  if (wait_for_servers == ros::WallDuration())
+    timeout_for_servers = ros::WallTime();  // wait for ever
+  double allotted_time = wait_for_servers.toSec();
+
+  move_action_client_.reset(
```
Comment (davetcoleman) [inline moveit_ros/planning_interface/moveit_cpp/src/moveit.cpp]: remove these action clients

File: moveit_ros/planning_interface/moveit_cpp/CMakeLists.txt [hunk 8]
```diff
@@ -0,0 +1,12 @@
+set(MOVEIT_LIB_NAME moveit_cpp)
+
+add_library(${MOVEIT_LIB_NAME} src/moveit.cpp)
```
Comment (davetcoleman) [inline moveit_ros/planning_interface/moveit_cpp/CMakeLists.txt]: this file should be called moveit_cpp.cpp, right? to match the header file

**Review by davetcoleman — COMMENTED** [review]
Gave it a quick review

**Comment by davetcoleman** [comment]
Rather than opening this PR against Simon's, I'd rather see the whole feature opened to main moveit repo

**Comment by henningkayser** [comment]
Closing in favor of a new PR against the main repo

**Pull request closed without merging**